---
title: "Using PublicDataNorm package for dumies"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Public-data-norm-Vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```


*Quick protocol to build a database from public Omics data.*

`PublicDataNorm` provides R functions to build a Metaobject (`AddExpressionMatrix()`) with Expression datasets, Patients and Samples clinical, pathological and so on annotations and Gene/Probes annotation. Then help to "Clean" public data (`CleaningClinic()`)into a customizable references for Patients and Samples annotations (`LexicData()`).

The database is built into a folder tree as follow:\n

* "Parent dir" of your choice.\n
  * 01RawData\n
  * 02Processes\n
  * 03References\n
  * 04VerifiedDataSet\n

Public genes expression dataset form The Cancer Genome Atlas (TCGA) can be retrive with the incorporation of `TCGAbiolinks` R apckage, incorporated in `PublicDataNorm` functions.

Export from Metaobject can be in CSV or TSV format, then compressed in gz using `gzip` package for gene expression data.

Along building and exporting the cleaned fetched public data, a summary of an exported project can be created into a Database summary `.txt` file. 

## Installation and loading package
```{r chunk, eval=c(3), message=F, warning=F, include=TRUE}
#Download zip file from : 
download.file("https://github.com/JRAnalytics/PublicDataNorm/tree/V2",method = "wget")
library(PublicDataNorm)
```



## Data base building steps

### Create dataBase directory and project directories

```{r DataBaseCreation}
project <- "PANC.RNAseq"

  Parent.Dir.path <- "./Parent" # Change parent path matching your own Dir
  
  #if it is the first time that you create the database,
  #list.files(Parent.Dir.path) will display zero character.
  list.files(Parent.Dir.path) 
```  
  
  DirCheck function will create Parent subdirectories\n
  * 01RawData\n
  * 02Processes\n
  * 03References\n
  * 04VerifiedDataSet\n

```{r DataBaseCreation2}
  #building data base in Parent dir, with Project sub directories.
  list.files.path <- DirCheck(project,Parent.Dir.path)
```




Filepaths are stock in list.files.path object. It will be usefull to keep path for subsequent package functions.
```{r DataBaseCreation3}
  list.files.path
```


## Building Meta object

### A-RNAseq projects

#### 1- Creating Meta object 

The first step is to create a meta object with en expression matrix saved in the project directory '~/parent/01RawData/Project'
```{r RNAseq, eval=FALSE}
Metadata <- AddExpressionMatrixRNAseq(Metadata = NULL,
                            Export = T,
                            Raw.file.path = list.files.path$Project.RawData,
                            name.local.file ="PANC.rawcount.csv",
                            name = "Raw.count")

```

If a Metadata object already exist, you can add another expression matrix to the meta object.
```{r RNAseq2, eval=FALSE}
Metadata <-AddExpressionMatrixRNAseq(Metadata = Metadata,
                                 Export = T,# True pour Export
                                 Raw.file.path = list.files.path$Project.RawData,
                                 name.local.file ="PANC.NormCount.csv",
                                 name = "Normalized.count")
```

Also, if a sub-object in Metadata has to be replaced, you can set 'force.replace' parameter to T.
```{r RNAseq3, eval=FALSE}
Metadata <- AddExpressionMatrixRNAseq(Metadata = Metadata,
                                 Export = T,
                                 Raw.file.path = list.files.path$Project.RawData,
                                 name.local.file ="PANC.rawcount2.csv",
                                 name = "Raw.count",
                                 force.replace = T # replace Metadata$PANC.rawcount.csv with the same name 
                                 )
```


#### 2- Adding clinical data

Patients clinical data or Samples characteristics can be added to the Meta object through AddClinicFromFile() function from a text file (.txt or .csv) in the project's raw.data file.

In this case, the loaded file is a Samples annotation file, so we set type by "Samples". If it was a Patients clincal file, it has to be set to "Patients". It is mandatory for cleaning this file in the next steps. 

```{r RNAseq.Clin1, eval=FALSE}
Metadata = AddClinicFromFile(Metadata = Metadata,
                         Raw.file.path = list.files.path$Project.RawData,
                         name.local.file ="PANC.samplesAnnotation.csv",
                         name = "Pheno",
                         type = "Samples")
```

Of course you can replace the Metadata sub-object "Pheno" by setting   force.replace = T. default is FALSE.

```{r RNAseq.Clin1b, eval=FALSE}
Metadata = AddClinicFromFile(Metadata = Metadata,
                         Raw.file.path = list.files.path$Project.RawData,
                         name.local.file ="PANC.samplesAnnotation2.csv",
                         name = "Pheno",
                         type = "Samples",
                         force.replace = T)
```

By default, this file are considered as raw Patients clinical/Samples annotation files and will not be exported at the final steps of cleaning. if you want to save the raw files in the project directory in 04VerifiedDataSet dir, you can set "Export = T" (default F).

```{r RNAseq.Clin2, eval=FALSE}
Metadata = AddClinicFromFile(Metadata = Metadata,
                         Raw.file.path = list.files.path$Project.RawData,
                         name.local.file ="PANC.samplesAnnotation.csv",
                         name = "Pheno",
                         type = "Samples",
                         Export =  T) #Default is F

```



```{r RNAseq.Clin3, eval=FALSE}
Metadata = AddClinicFromFile(Metadata = Metadata,
                         Raw.file.path = list.files.path$Project.RawData,
                         name.local.file = c( "PANC.samplesAnnotation.csv", "PANC.Clinic.all.csv"),
                         mergeBy = "id_cit", #mandatory if name.local.file >1 string
                         join = "full_join", #mandatory if name.local.file >1 string
                         name = "Pheno",
                         type = "Samples")


```


